fortuna/prob_model/posterior/sgmcmc/sgmcmc_diagnostic.py

```diff
@@ -0,0 +1,134 @@
+from typing import Optional, List
+from optax._src.base import PyTree
+
+import jax.numpy as jnp
+from jax import jit, lax, vmap
+from jax.flatten_util import ravel_pytree
+
+
+@jit
+def kernel_stein_discrepancy_imq(
+    samples: List[PyTree],
+    grads: List[PyTree],
+    c: float = 1.0,
+    beta: float = -0.5,
+) -> float:
+    """Kernel Stein Discrepancy with the Inverse Multiquadric (IMQ) kernel.
+
+    See `Gorham J. and Mackey L., 2017 <https://proceedings.mlr.press/v70/gorham17a/gorham17a.pdf>`_ for more details.
+
+    Parameters
+    ----------
+        samples: List[PyTree]
+            The list of `PyTree`, each representing an MCMC sample.
+        grads: List[PyTree]
+            The list of the corresponding density gradients.
+        c: float
+            :math:`c > 0` kernel bias hyperparameter.
+        beta: float
+            :math:`beta < 0` kernel exponent hyperparameter.
+
+    Returns
+    -------
+        ksd_img: float
+            The kernel Stein discrepancy value.
+    """
+    samples = ravel_pytree(samples)[0].reshape(len(samples), -1)
+    grads = ravel_pytree(grads)[0].reshape(len(grads), -1)
+
+    def _k_0(param1, param2, grad1, grad2, c, beta):
+        dim = param1.shape[0]
+        diff = param1 - param2
+        base = c**2 + jnp.dot(diff, diff)
+        kern = jnp.dot(grad1, grad2) * base**beta
+        kern += -2 * beta * jnp.dot(grad1, diff) * base ** (beta - 1)
+        kern += 2 * beta * jnp.dot(grad2, diff) * base ** (beta - 1)
```

Comment:
It looks like the `base` always goes at the denominator, since `beta` is negative. For example, with the default `beta=-0.5`, we have `... / base ** (3/2)`.

Just want to check, in case this is not the intended behaviour.

Reply:
It's a re-implementation of the reference code and I hope it's correct: https://github.com/jgorham/SteinDiscrepancy.jl/blob/master/src/kernels/SteinInverseMultiquadricKernel.jl (notice that they [flipped the sign in the parametrization](https://github.com/jgorham/SteinDiscrepancy.jl/blob/master/src/kernels/SteinInverseMultiquadricKernel.jl#L8)).